File: java/src/main/java/it/giannibombelli/wsc2026/booking/application/services/PaymentResultOutcome.java
package it.giannibombelli.wsc2026.booking.application.services;

import it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands;
import it.giannibombelli.wsc2026.booking.application.usecases.BookingConfirming;
import it.giannibombelli.wsc2026.booking.application.usecases.BookingRejecting;
import it.giannibombelli.wsc2026.booking.domain.policies.PaymentPolicy;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;

import static it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands.ConfirmBooking;
import static it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands.RejectBooking;
import static java.util.Objects.requireNonNull;


public class PaymentResultOutcome {
    private final PaymentPolicy policy;
    private final BookingConfirming confirmation;
    private final BookingRejecting rejection;

    public PaymentResultOutcome(PaymentPolicy policy, BookingConfirming confirmation, BookingRejecting rejection) {
        this.policy = requireNonNull(policy);
        this.confirmation = requireNonNull(confirmation);
        this.rejection = requireNonNull(rejection);
    }

    public void handlePaymentResults(PaymentResultEvents event) {
        BookingConfirmationCommands cmd = policy.evaluate(event);
        if (cmd == null) {
            return;
        }
        switch (cmd) {
            case ConfirmBooking confirmBooking -> confirmation.invoke(confirmBooking);
            case RejectBooking rejectBooking -> rejection.invoke(rejectBooking);
        }
    }
}

File: typescript/src/booking/application/services/paymentResultOutcome.ts
// Servizio applicativo che reagisce agli esiti del Payment BC confermando/rifiutando la prenotazione.


import type { PaymentResultEvent } from '@/payment/domain/events/paymentResultEvents.js';
import { match } from 'ts-pattern';
import { PaymentPolicy } from '../../application/policies/paymentPolicy.js';
import { BookingConfirming } from '../usecases/bookingConfirming.js';
import { BookingRejecting } from '../usecases/bookingRejecting.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class PaymentResultOutcome {
  private readonly _policy: PaymentPolicy;
  private readonly _confirmation: BookingConfirming;
  private readonly _rejection: BookingRejecting;

  constructor(policy: PaymentPolicy, confirmation: BookingConfirming, rejection: BookingRejecting) {
    requireDependency(policy, "policy");
    requireDependency(confirmation, "confirmation");
    requireDependency(rejection, "rejection");
    this._policy = policy;
    this._confirmation = confirmation;
    this._rejection = rejection;
  }

  handlePaymentResults(event: PaymentResultEvent): void {
    const command = this._policy.evaluate(event);
    if (command === null) {
      return;
    }

    match(command)
      .with({ kind: 'ConfirmBooking' }, (confirmBooking) => this._confirmation.invoke(confirmBooking))
      .with({ kind: 'RejectBooking' }, (rejectBooking) => this._rejection.invoke(rejectBooking))
      .exhaustive();
  }
}
